dataservice/api/demographic/resources.py

```diff
@@ -0,0 +1,115 @@
+from flask import (
+    abort,
+    request
+)
+from flask.views import MethodView
+from sqlalchemy.exc import IntegrityError
+from sqlalchemy.orm.exc import NoResultFound
+from marshmallow import ValidationError
+
+from dataservice.extensions import db
+from dataservice.api.errors import handle_integrity_error
+from dataservice.api.demographic.models import Demographic
+from dataservice.api.demographic.schemas import DemographicSchema
+
+
+class DemographicAPI(MethodView):
+    """
+    Demographic REST API
+    """
+
+    def get(self, kf_id):
+        """
+        Get a demographic by id or get all demographics
+
+        Get a demographic by Kids First id or get all demographics if
+        Kids First id is None
+        """
+        # Get all
+        if kf_id is None:
+            d = Demographic.query.all()
+            return DemographicSchema(many=True).jsonify(d)
+        # Get one
+        else:
+            try:
+                d = Demographic.query.filter_by(kf_id=kf_id).one()
+            # Not found in database
+            except NoResultFound:
+                abort(404, 'could not find {} `{}`'
+                      .format('demographic', kf_id))
+            return DemographicSchema().jsonify(d)
+
+    def post(self):
+        """
+        Create a new demographic
+        """
+        body = request.json
+        try:
+            # Deserialize
+            d = DemographicSchema(strict=True).load(body).data
+            # Add to and save in database
+            db.session.add(d)
+            db.session.commit()
+        # Request body not valid
+        except ValidationError as e:
+            abort(400, 'could not create demographic: {}'.format(e.messages))
+        # Database error
+        except IntegrityError as e:
+            db.session.rollback()
+            context = {'method': 'create', 'entity': 'demographic',
+                       'ref_entity': 'participant', 'exception': e}
+            abort(400, handle_integrity_error(**context))
+
+        return DemographicSchema(201, 'demographic {} created'
+                                 .format(d.kf_id)).jsonify(d), 201
+
+    def put(self, kf_id):
+        """
+        Update existing demographic
+
+        Update an existing demographic given a Kids First id
+        """
+        body = request.json
+        try:
+            # Check if demographic exists
+            d1 = Demographic.query.filter_by(kf_id=kf_id).one()
+            # Deserialize
```

Comment:
This can be done more dynamically/better in the future. Tried using sqlalchemy session.merge with the loaded obj from marshmallow but this achieves patch/partial update behavior. To do a full replace, would first have to iterate over the d1's columns and relationships and set them back to defaults and then perform merge with the deserialized object from marshmallow. This isn't trivial so will look into this in the near future.